packages/std/src/math/decimal.rs

```diff
@@ -129,24 +128,24 @@ impl FromStr for Decimal {
 
         let whole_part = parts_iter.next().unwrap(); // split always returns at least one element
         let whole = whole_part
-            .parse::<u128>()
+            .parse::<Uint128>()
             .map_err(|_| StdError::generic_err("Error parsing whole"))?;
         let mut atomics = whole
-            .checked_mul(DECIMAL_FRACTIONAL)
-            .ok_or_else(|| StdError::generic_err("Value too big"))?;
+            .checked_mul(Self::DECIMAL_FRACTIONAL)
+            .map_err(|_| StdError::generic_err("Value too big"))?;
 
         if let Some(fractional_part) = parts_iter.next() {
             let fractional = fractional_part
-                .parse::<u128>()
+                .parse::<Uint128>()
                 .map_err(|_| StdError::generic_err("Error parsing fractional"))?;
             let exp = (18usize.checked_sub(fractional_part.len())).ok_or_else(|| {
                 StdError::generic_err("Cannot parse more than 18 fractional digits")
             })?;
             debug_assert!(exp <= 18);
-            let fractional_factor = 10u128.pow(exp as u32);
+            let fractional_factor = Uint128::from(10u128.pow(exp as u32));
             atomics = atomics
-                .checked_add(fractional * fractional_factor)
-                .ok_or_else(|| StdError::generic_err("Value too big"))?;
+                .checked_add(fractional.checked_mul(fractional_factor)?)
```

Comment:
The inner multiplication here cannot overflow because fractional <= 999999999999999999 and fractional_factor <= 10**18. Can we make this clear in code?